spring-batch-core/src/main/java/org/springframework/batch/core/job/flow/support/SimpleFlow.java

```diff
@@ -260,6 +258,19 @@ protected boolean isFlowContinued(State state, FlowExecutionStatus status, StepE
 		return continued;
 	}
 
+	/**
+	 * To prevent race condition when two or more threads initialize transitions concurrently.
+	 */
+	private void initializeTransitionsIfNotInitialized() {
+		if (startState == null) {
+			synchronized (this) {
+				if (startState == null) {
+					initializeTransitions();
+				}
+			}
+		}
+	}
+

```

Comment:
I think the first way is better. This is because the modified method continues to lock due to synchronized keyword even after the instance has already been created.